src/test/java/io/cdap/plugin/WindowsAggregationUtilTest.java

```diff
@@ -190,4 +194,20 @@ public void testInvalidParameter() {
                                                                           config, schema, schema);
     }
   }
+
+  @Test
+  public void testWindowSpec() throws NoSuchFieldException, IllegalAccessException {
+    WindowSpec spec = Window.partitionBy(WindowsAggregationUtil.getPartitionsColumns(config.getPartitionFields()))
+                            .orderBy(WindowsAggregationUtil.getPartitionOrderColumns(config.getPartitionOrder()));
+    spec = spec.rangeBetween(config.getFrameDefinitionPrecedingBound(),
```

Comment:
I *think* I found the issue.

As per the [javadoc](https://spark.apache.org/docs/2.3.0/api/java/org/apache/spark/sql/expressions/WindowSpec.html#rangeBetween-long-long-), the argument to the `rangeBetween()` function is a number of rows offset from the current row, where 0 is the current row, a positive number represents a row following the current row and a negative number represents a row preceding the current row. Therefore, if we want to have a window with rows preceding the current row, the row number should be **negative**.

Spark source code where this is handled is present in [windowExpressions.scala](https://github.com/apache/spark/blob/master/sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/windowExpressions.scala#L271).

TL;DR: Use a negative number for `PRECEDING`.

Reply:
Yes .For preceding negative number is taken.IF it is positive number it gives the current response.